cpp/src/parquet/variant.cc

```diff
@@ -0,0 +1,779 @@
+      return "Uuid";
+    default:
+      return "Unknown";
+  }
+}
+
+std::string VariantTypeToString(VariantType type) {
+  switch (type) {
+    case VariantType::Object:
+      return "Object";
+    case VariantType::Array:
+      return "Array";
+    case VariantType::Null:
+      return "Null";
+    case VariantType::Boolean:
+      return "Boolean";
+    case VariantType::Int8:
+      return "Int8";
+    case VariantType::Int16:
+      return "Int16";
+    case VariantType::Int32:
+      return "Int32";
+    case VariantType::Int64:
+      return "Int64";
+    case VariantType::String:
+      return "String";
+    case VariantType::Double:
+      return "Double";
+    case VariantType::Decimal4:
+      return "Decimal4";
+    case VariantType::Decimal8:
+      return "Decimal8";
+    case VariantType::Decimal16:
+      return "Decimal16";
+    case VariantType::Date:
+      return "Date";
+    case VariantType::TimestampMicrosTz:
+      return "TimestampMicrosTz";
+    case VariantType::TimestampMicrosNtz:
+      return "TimestampMicrosNtz";
+    case VariantType::Float:
+      return "Float";
+    case VariantType::Binary:
+      return "Binary";
+    case VariantType::Time:
+      return "Time";
+    case VariantType::TimestampNanosTz:
+      return "TimestampNanosTz";
+    case VariantType::TimestampNanosNtz:
+      return "TimestampNanosNtz";
+    case VariantType::Uuid:
+      return "Uuid";
+    default:
+      return "Unknown";
+  }
+}
+
+inline uint32_t readLittleEndianU32(const void* from, uint8_t size) {
+  ARROW_DCHECK_LE(size, 4);
+  ARROW_DCHECK_GE(size, 1);
+
+  uint32_t result = 0;
+  memcpy(&result, from, size);
+  return ::arrow::bit_util::FromLittleEndian(result);
+}
+
+VariantMetadata::VariantMetadata(std::string_view metadata) : metadata_(metadata) {
+  if (metadata.size() < kHeaderSizeBytes + kMinimalOffsetSizeBytes * 2) {
+    // Empty metadata is at least 3 bytes: version, dictionarySize and
+    // at least one offset.
+    throw ParquetException("Invalid Variant metadata: too short: size=" +
+                           std::to_string(metadata.size()));
+  }
+  if (version() != kSupportedVersion) {
+    // Currently we only supports version 1.
+    throw ParquetException("Unsupported Variant metadata version: " +
+                           std::to_string(version()));
+  }
+  uint8_t offset_sz = offset_size();
+  if (offset_sz < kMinimalOffsetSizeBytes || offset_sz > kMaximumOffsetSizeBytes) {
+    throw ParquetException("Invalid Variant metadata: invalid offset size: " +
+                           std::to_string(offset_sz));
+  }
+  dictionary_size_ = loadDictionarySize(metadata, offset_sz);
+  if (kHeaderSizeBytes + (dictionary_size_ + 2) * offset_sz > metadata_.size()) {
+    throw ParquetException(
+        "Invalid Variant metadata: offset out of range: " +
+        std::to_string((dictionary_size_ + kHeaderSizeBytes) * offset_sz) + " > " +
+        std::to_string(metadata_.size()));
+  }
```

Comment:
In metadata I delayed the validating offsets to getter, it can also put here to validate